upf/upf_pfcp_api.c

```diff
@@ -803,9 +809,210 @@ lookup_nwi (u8 * name)
   return pool_elt_at_index (gtm->nwis, p[0]);
 }
 
+static bool
+validate_teid (u32 teid)
+{
+  return ((teid != 0) && (teid != 0xffffffff));
+}
+
+static u32
+generate_teid (u32 start, u32 mask)
+{
+  u8 ret_cnt = 0;
+  u32 c = 0;
+
+retry:
+
+  if (ret_cnt > 1)
+    return 0;
+
+  c = random_u32(&start) & (UINT32_MAX ^ mask);
+  if (!validate_teid(c))
+    {
+      //we are doing single try to regenerate, making next
+      //based on generated value
+      ret_cnt += 1;
+      goto retry;
+    }
+
+  return c;
+}
+
+static int
+teid_v4_lookup (upf_main_t *gtm, u32 teid, upf_upip_res_t *res)
+{
+  clib_bihash_kv_8_8_t kv, value;
+  gtpu4_tunnel_key_t key4;
+
+  key4.dst = res->ip4.as_u32;
+  key4.teid = teid;
+
+  kv.key = key4.as_u64;
+
+  if (PREDICT_FALSE
+	(clib_bihash_search_8_8 (&gtm->v4_tunnel_by_key, &kv, &value)))
+            return UPF_GTPU_ERROR_NO_SUCH_TUNNEL;
+
+  return 0;
+}
+
+
+static int
+teid_v6_lookup (upf_main_t *gtm, u32 teid, upf_upip_res_t *res)
+{
+  clib_bihash_kv_24_8_t kv, value;
+
+  kv.key[0] = res->ip6.as_u64[0];
+  kv.key[1] = res->ip6.as_u64[1];
+  kv.key[2] = teid;
+
+  if (PREDICT_FALSE
+	(clib_bihash_search_24_8 (&gtm->v6_tunnel_by_key, &kv, &value)))
+            return UPF_GTPU_ERROR_NO_SUCH_TUNNEL;
+
+  return 0;
+}
+
+static u32
+process_teid_generation (upf_main_t *gtm, u8 chid, u32 flags,
+		  upf_upip_res_t *res, upf_session_t *sx)
+{
+  u8 retry_cnt = 0;
+  bool added = false;
+  u32 teid = 0;
+
+  teid = generate_teid(gtm->rand_base, res->mask);
+
+  do
+    {
+      if (flags & F_TEID_V4)
+	{
+	  if (!(teid_v4_lookup(gtm, teid, res)))
+	    {
+	      teid = generate_teid(teid + ( retry_cnt + 1 ), res->mask);
+	    }
+	  else
+	    {
+	      added = true;
+	      break;
+	    }
+	}
+      if ((flags & F_TEID_V6) && !added)
+        {
+          if (!(teid_v6_lookup(gtm, teid, res)))
+            {
+              teid = generate_teid(teid + ( retry_cnt + 1 ), res->mask);
+            }
+          else
+            {
+              added = true;
+              break;
+            }
+        }
+      retry_cnt++;
+    }
+  while (retry_cnt < 10);
+
+  if (!added)
+    {
+      //can't generate unique for given UPIP resource
+      return 0;
+    }
+
+  /* generate_teid can 0 if for 2 consecutive tries we will
+     to generate teid different from 0 and 0xffffffff
+  */
+  if (!validate_teid(teid))
+    return 0;
+
+  gtm->rand_base = teid;
+  if ((flags & F_TEID_CHID))
+    {
+      u32 *n;
+      n = sparse_vec_validate (sx->teid_by_chid, chid);
+      n[0] = teid;
+    }
+
+  return teid;
+}
```

Comment:
too complicated, untested

```suggestion
static u32
process_teid_generation (upf_main_t *gtm, u8 chid, u32 flags,
		  upf_upip_res_t *res, upf_session_t *sx)
{
  u8 retry_cnt = 10;
  u32 teid = 0;
  bool ok;


  do
    {
      teid = generate_teid(gtm->rand_base, res->mask);
      gtm->rand_base = teid + retry_cnt;

      if (!validate_teid(teid))
	{
	  retry_cnt--;
	  continue;
	}

      ok =
	(!(flags & F_TEID_V4) || teid_v4_lookup(gtm, teid, res) == 0) &&
	(!(flags & F_TEID_V6) || teid_v6_lookup(gtm, teid, res) == 0);

      if (ok)
	break;

    }
  while (retry_cnt > 0);

  if (!ok)
    /* can't generate unique for given UPIP resource */
    return 0;

  if ((flags & F_TEID_CHID))
    {
      u32 *n = sparse_vec_validate (sx->teid_by_chid, chid);
      n[0] = teid;
    }

  return teid;
}
```